File: wiz-event-mall-java/src/test/java/com/example/wizeventmalljava/api/event/EventServiceTest.java
package com.example.wizeventmalljava.api.event;

import static org.assertj.core.api.Assertions.*;

import java.time.LocalDateTime;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;

import com.example.wizeventmalljava.domain.event.entity.Event;
import com.example.wizeventmalljava.domain.event.repository.EventRepository;
import com.example.wizeventmalljava.domain.event.repository.EventWinnerRepository;
import com.example.wizeventmalljava.domain.member.entity.Member;
import com.example.wizeventmalljava.domain.member.repository.MemberRepository;

@SpringBootTest
@ActiveProfiles("test")
public class EventServiceTest {

	private final static int THREAD_COUNT = 500;

	@Autowired
	private EventRepository eventRepository;

	@Autowired
	private EventWinnerRepository eventWinnerRepository;

	@Autowired
	private MemberRepository memberRepository;

	@Autowired
	private EventService eventService;

	@Test
	@DisplayName(value = "Participate Event - 동시성 테스트")
	public void concurrencyTest() throws InterruptedException {
		// GIVEN
		var executor = Executors.newFixedThreadPool(THREAD_COUNT);
		var barrier = new CyclicBarrier(THREAD_COUNT);

		for (int i = 0; i < THREAD_COUNT; i++) {
			memberRepository.save(
				Member.builder().email("slolee@naver.com").password("TEST!@#$").nickname("wiz" + i).build());
		}
		Event event = eventRepository.save(Event.builder().
			name("Test Event")
			.startAt(LocalDateTime.now().minusHours(1))
			.winnerCount(100L)
			.build());

		// WHEN
		for (int i = 0; i < THREAD_COUNT; i++) {
			int memberId = i + 1;
			executor.execute(() -> {
				try {
					barrier.await(); // 모든 스레드가 동시에 시작되도록 대기
					eventService.participateWithLock(event.getId(), (long) memberId);
				} catch (Exception e) {
					throw new RuntimeException(e);
				}
			});
		}
		executor.shutdown();
		executor.awaitTermination(5, TimeUnit.SECONDS);

		// THEN
		assertThat(eventWinnerRepository.findAll().size()).isEqualTo(100);
	}
}

File: wiz-event-mall/src/test/kotlin/com/example/wizeventmall/api/event/EventServiceTest.kt
package com.example.wizeventmall.api.event

import com.example.wizeventmall.domain.event.entity.Event
import com.example.wizeventmall.domain.event.repository.EventRepository
import com.example.wizeventmall.domain.event.repository.EventWinnerRepository
import com.example.wizeventmall.domain.member.entity.Member
import com.example.wizeventmall.domain.member.repository.MemberRepository
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import java.time.LocalDateTime
import java.util.concurrent.CyclicBarrier
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

@SpringBootTest
@ActiveProfiles("test")
class EventServiceTest @Autowired constructor(
    private val eventService: EventService,

    private val eventRepository: EventRepository,
    private val eventWinnerRepository: EventWinnerRepository,
    private val memberRepository: MemberRepository
) {

    @Test
    fun `Participate Event - 동시성 테스트`() {
        // GIVEN
        val executor = Executors.newFixedThreadPool(THREAD_COUNT)
        val barrier = CyclicBarrier(THREAD_COUNT)

        // 1. Thread 수만큼 사용자 생성
        repeat(THREAD_COUNT) {
            memberRepository.save(Member(email = "slolee@naver.com", password = "TEST!@#$", nickname = "wiz$it"))
        }
        // 2. Event 정보 생성
        val event = Event(
            name = "Test Event",
            startAt = LocalDateTime.now().minusHours(1),
            winnerCount = 10
        ).let { eventRepository.save(it) }

        // WHEN
        repeat(THREAD_COUNT) {
            executor.execute {
                barrier.await()
                eventService.participate(eventId = event.id!!, memberId = it.toLong() + 1)
            }
        }
        executor.awaitTermination(5, TimeUnit.SECONDS)

        // THEN
        eventWinnerRepository.findAll().size shouldBe 10
    }

    companion object {
        private const val THREAD_COUNT = 100
    }
}